utils/check_copies.py

```diff
@@ -181,22 +387,91 @@ def find_code_in_transformers(object_name: str, base_path: str = None) -> str:
             line_index < len(lines) and re.search(rf"^{indent}(class|def)\s+{name}(\(|\:)", lines[line_index]) is None
         ):
             line_index += 1
+        # find the target specified in the current level in `parts` -> increase `indent` so we can search the next
         indent += "    "
+        # the index of the first line in the (currently found) block *body*
         line_index += 1
 
     if line_index >= len(lines):
         raise ValueError(f" {object_name} does not match any function or class in {module}.")
 
+    # `indent` is already one level deeper than the (found) class/func block's definition header
+
     # We found the beginning of the class / func, now let's find the end (when the indent diminishes).
+    # `start_index` is the index of the class/func block's definition header
     start_index = line_index - 1
-    while line_index < len(lines) and _should_continue(lines[line_index], indent):
-        line_index += 1
-    # Clean up empty lines at the end (if any).
-    while len(lines[line_index - 1]) <= 1:
-        line_index -= 1
+    end_index = find_block_end(lines, start_index, len(indent))
+
+    code = "".join(lines[start_index:end_index])
+    return (code, (lines, start_index, end_index)) if return_indices else code
+
 
-    code_lines = lines[start_index:line_index]
-    return "".join(code_lines)
+def replace_code(code: str, replace_pattern: str) -> str:
+    """Replace `code` by a pattern of the form `with X1->X2,Y1->Y2,Z1->Z2`.
+
+    Args:
+        code (`str`): The code to be modified.
+        replace_pattern (`str`): The pattern used to modify `code`.
+
+    Returns:
+        `str`: The modified code.
+    """
+    if len(replace_pattern) > 0:
+        patterns = replace_pattern.replace("with", "").split(",")
+        patterns = [_re_replace_pattern.search(p) for p in patterns]
+        for pattern in patterns:
+            if pattern is None:
+                continue
+            obj1, obj2, option = pattern.groups()
+            code = re.sub(obj1, obj2, code)
+            if option.strip() == "all-casing":
+                code = re.sub(obj1.lower(), obj2.lower(), code)
+                code = re.sub(obj1.upper(), obj2.upper(), code)
+
+    return code
+
+
+def find_code_and_splits(object_name: str, base_path: str, buf: dict = None):
+    """Find the code of an object (specified by `object_name`) and split it into blocks.
+
+    Args:
+        object_name (`str`):
+            The name of the object, e.g. `transformers.models.bert.modeling_bert.BertAttention` or
+            `tests.models.llama.test_modeling_llama.LlamaModelTest.test_config`.
+        base_path (`str`):
+            The path to the base directory within which the search will be performed. It could be either
+            `TRANSFORMERS_PATH` or `MODEL_TEST_PATH`.
+        buf (`dict`, *optional*):
+            The buffer used to store the previous results in order to speed up the process.
+
```

Comment:
how much speedup do we have? 🤗

Reply:
thinking if the buffer could store already formatted code would be nice as well!

Reply:
very little unfortunately, even before your work on using ruff. I can remove this part.

> the buffer could store already formatted code

It's possible to use another (different) buffer to do it, but not this existing one.